staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go

```diff
@@ -0,0 +1,1459 @@
+	// before write timeout occurs:
+	//  - a) the handler writes a few bytes so there is some content to be flushed later
+	//  - b) write deadline is set to 100ms
+	//  - c) wait until the write deadline elapses
+	//  - d) the handler invokes FlushError
+	// timeout occurs after the handler writes to but has not flushed the
+	// ResponseWriter object yet, we expect the following:
+	//  - e) client: expected to receive an error from the server
+	/// - f) server: FlushError method of the ResponseWriter object should
+	//       return an "i/o timeout" error immediately
+	t.Parallel()
+
+	const deadline = 100 * time.Millisecond
+	tests := []struct {
+		protoMajor         int
+		waiter             waiter
+		clientErrVerifier  verifier
+		handlerErrVerifier verifier
+	}{
+		{
+			protoMajor: 1,
+			// write timeout is set to 100ms, a wait of 5s should
+			// be enough to withstand flakes in CI.
+			waiter:             &waitWithDuration{after: 5 * time.Second},
+			clientErrVerifier:  wantContains{contains: "local error: tls: bad record MAC"},
+			handlerErrVerifier: wantError{err: os.ErrDeadlineExceeded},
+		},
+		{
+			protoMajor:         2,
+			waiter:             &waitWithChannelClose{after: make(chan time.Time)},
+			clientErrVerifier:  wantContains{"stream error: stream ID 1; INTERNAL_ERROR; received from peer"},
+			handlerErrVerifier: wantError{err: os.ErrDeadlineExceeded},
+		},
+	}
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("HTTP/%d", test.protoMajor), func(t *testing.T) {
+			handlerErrCh := make(chanErr, 1)
+			server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+				defer close(handlerErrCh)
+				if req.ProtoMajor != test.protoMajor {
+					t.Errorf("server: expected ProtoMajor: %d, but got: %s", test.protoMajor, req.Proto)
+					return
+				}
+				flusher, ok := w.(interface{ FlushError() error })
+				if !ok {
+					t.Errorf("server: expected ResponseWriter object to implement FlushError")
+					return
+				}
+
+				// a) write a few bytes so the internal buffer
+				// is not full, don't flush yet
+				if _, err := w.Write([]byte("hello")); err != nil {
+					t.Errorf("server: unexpected error from Write: %v", err)
+					return
+				}
+
+				// b) set write deadline for this request
+				ctrl := http.NewResponseController(w)
+				if err := ctrl.SetWriteDeadline(time.Now().Add(deadline)); err != nil {
+					t.Errorf("server: unexpected error from SetWriteDeadline: %v", err)
+					return
+				}
+
+				// c) wait until the write deadline exceeds
+				<-test.waiter.wait()
+
+				// d) flush the payload in the buffer
+				if err := flusher.FlushError(); err != nil {
+					handlerErrCh <- err
+				}
+			}))
+
+			defer server.Close()
+			if test.protoMajor == 2 {
+				server.EnableHTTP2 = true
+			}
+			server.StartTLS()
+
+			func() {
+				defer test.waiter.close()
+
+				_, err := server.Client().Get(server.URL)
```

Comment:
What surprises is that you get an error, but there has been data written before. So wouldn't I expect a body? And then wouldn't I have to close that body? I.e. this is leaking.